File: mathai/expand.py
from .base import *
from .simplify import simplify
import itertools

def expand_nc(expr, label="f_mul"):

    if expr.name not in {"f_add", label, "f_pow"}:
        return expr

    expr.children = [expand_nc(c, label) for c in expr.children]

    if expr.name == "f_pow":
        base, exp = expr.children
        n = frac(exp)
        if n and n.denominator == 1 and n.numerator > 1:
            factors = [base] * n.numerator
            return expand_nc(TreeNode(label, factors), label)
        return expr

    if expr.name == "f_add":
        out = []
        for c in expr.children:
            if c.name == "f_add":
                out.extend(c.children)
            else:
                out.append(c)
        return TreeNode("f_add", out)

    if expr.name == label:
        factors = []

        for c in expr.children:
            if c.name == label:
                factors.extend(c.children)
            else:
                factors.append(c)

        for i, f in enumerate(factors):
            if f.name == "f_add":
                left  = factors[:i]
                right = factors[i+1:]

                terms = []
                for term in f.children:
                    new_factors = left + [term] + right
                    terms.append(
                        expand_nc(TreeNode(label, new_factors), label)
                    )

                return TreeNode("f_add", terms)

        return TreeNode(label, factors)

def expand2(eq, over="*"):
    over = {"@": "f_wmul", ".":"f_dot", "*":"f_mul"}[over]
    return expand_nc(eq, over)
def expand(eq, over="*"):
    eq = expand2(eq, over)
    return TreeNode(eq.name, [expand(child, over) for child in eq.children])

File: haskell/Expand.hs
module Expand where

import Base
import Data.Maybe (fromMaybe)
import Data.Ratio (numerator, denominator)

------------------------------------------------------------
-- expand_nc
------------------------------------------------------------

expandNC :: TreeNode -> String -> TreeNode
expandNC expr label
  | name expr `notElem` ["f_add", label, "f_pow"] = expr
  | otherwise =
      case name expr of

        --------------------------------------------------
        -- f_pow
        --------------------------------------------------
        "f_pow" ->
          let [base, expn] = map (`expandNC` label) (children expr)
          in case frac expn of
               Just r
                 | denominator r == 1
                 , numerator r > 1 ->
                     let n = fromIntegral (numerator r)
                         factors = replicate n base
                     in expandNC (TreeNode label factors) label
               _ -> TreeNode "f_pow" [base, expn]

        --------------------------------------------------
        -- f_add (flatten)
        --------------------------------------------------
        "f_add" ->
          let cs = map (`expandNC` label) (children expr)
              out = concatMap
                      (\c -> if name c == "f_add"
                             then children c
                             else [c])
                      cs
          in TreeNode "f_add" out

        --------------------------------------------------
        -- multiplication-like label
        --------------------------------------------------
        _ | name expr == label ->
            let cs = map (`expandNC` label) (children expr)
                factors =
                  concatMap
                    (\c -> if name c == label
                           then children c
                           else [c])
                    cs
            in distribute factors label

        _ -> expr


------------------------------------------------------------
-- distribute helper (core non-commutative expansion)
------------------------------------------------------------

distribute :: [TreeNode] -> String -> TreeNode
distribute factors label =
  case break ((== "f_add") . name) factors of
    (_, []) ->
      TreeNode label factors

    (left, addNode:right) ->
      let terms =
            [ expandNC (TreeNode label (left ++ [term] ++ right)) label
            | term <- children addNode
            ]
      in TreeNode "f_add" terms


------------------------------------------------------------
-- expand2
------------------------------------------------------------

expand2 :: TreeNode -> String -> TreeNode
expand2 eq over =
  let label = case over of
                "@" -> "f_wmul"
                "." -> "f_dot"
                "*" -> "f_mul"
                _   -> "f_mul"
  in expandNC eq label


------------------------------------------------------------
-- expand (recursive post-pass)
------------------------------------------------------------

expand :: TreeNode -> String -> TreeNode
expand eq over =
  let eq' = expand2 eq over
  in TreeNode (name eq') (map (\c -> expand c over) (children eq'))
